autonomi/src/client/data_types/chunk.rs

```diff
@@ -225,100 +228,82 @@ impl Client {
-                        .map_err(|err| (chunk, err));
-                    #[cfg(feature = "loud")]
-                    match res {
-                        Ok(_addr) => {
-                            println!(
-                                "({}/{}) Chunk stored at: {}",
-                                _i + 1,
-                                total_chunks,
-                                chunk.address().to_hex()
-                            );
-                        }
-                        Err((_chunk, ref err)) => {
-                            println!(
-                                "({}/{}) Chunk failed to be stored at: {} ({err})",
-                                _i + 1,
-                                total_chunks,
-                                chunk.address().to_hex()
-                            );
-                        }
+            upload_tasks.push(async move {
+                let res = self_clone
+                    .chunk_upload_with_payment(chunk, proof.clone(), *price)
+                    .await
+                    .inspect_err(|err| error!("Error uploading chunk {address:?} :{err:?}"))
+                    .map_err(|e| (chunk, e));
+                #[cfg(feature = "loud")]
+                match &res {
+                    Ok(_addr) => {
+                        println!(
+                            "({}/{}) Chunk stored at: {}",
+                            _i + 1,
+                            total_chunks,
+                            chunk.address().to_hex()
+                        );
                     }
-                    res
-                });
-            }
-            let uploads =
-                process_tasks_with_max_concurrency(upload_tasks, *CHUNK_UPLOAD_BATCH_SIZE).await;
-
-            // Check for errors.
-            let total_uploads = uploads.len();
-            let uploads_failed: Vec<_> = uploads.into_iter().filter_map(|up| up.err()).collect();
-            info!(
-                "Uploaded {} chunks out of {total_uploads}",
-                total_uploads - uploads_failed.len()
-            );
-
-            // All uploads succeeded.
-            if uploads_failed.is_empty() {
-                return vec![];
-            }
-
-            // Max retries reached.
-            if current_attempt > RETRY_ATTEMPTS {
-                return uploads_failed;
+                    Err((_, err)) => {
+                        println!(
+                            "({}/{}) Chunk failed to be stored at: {} ({err})",
+                            _i + 1,
+                            total_chunks,
+                            chunk.address().to_hex()
+                        );
+                    }
+                }
+                res
+            });
+        }
+        let uploads =
+            process_tasks_with_max_concurrency(upload_tasks, *CHUNK_UPLOAD_BATCH_SIZE).await;
+
+        // return errors
+        if uploads.iter().any(|res| res.is_err()) {
```

Comment:
can use `split` to avoid un-necessary iteration

Reply:
Although I agree that would be an optimization, here split would reduce readability. I'd prefer to avoid premature optimization here as I doubt this is the bottleneck in our code.